src/v2/git_url.rs

```diff
@@ -134,3 +172,39 @@ fn to_url_converts_git_urls_to_real_ones() {
         assert!(GitUrl::new(url).is_err());
     }
 }
+
+#[test]
+fn it_can_extract_its_repo_branch_and_subdir_parts() {
+    let urls = &[
+        "git://github.com/docker/docker",
+        "https://github.com/docker/docker.git",
+        "http://github.com/docker/docker.git",
+        "git@github.com:docker/docker.git",
+        "git@bitbucket.org:atlassianlabs/atlassian-docker.git",
+        "github.com/docker/docker.git",
+    ];
+
+    // Refs/folders specified as per:
+    // https://docs.docker.com/engine/reference/commandline/build/#git-repositories
+    for &url in urls {
+        let plain = GitUrl::new(url).unwrap();
+        assert_eq!(plain.repository(), url);
+        assert_eq!(plain.branch(), None);
+        assert_eq!(plain.subdirectory(), None);
+
+        let with_ref = GitUrl::new(format!("{}{}", url, "#mybranch")).unwrap();
+        assert_eq!(with_ref.repository(), url);
+        assert_eq!(with_ref.branch(), Some("mybranch".to_string()));
+        assert_eq!(with_ref.subdirectory(), None);
+
+        let with_subdir = GitUrl::new(format!("{}{}", url, "#:myfolder")).unwrap();
```

Comment:
This is great, but let's also add a test case for `format!("{}{}", url, ":myfolder"))`, specifying how that should be parsed.

Reply:
Looking at [the Docker docs](https://docs.docker.com/engine/reference/commandline/build/#git-repositories), I don't think that's a valid URL? Also, the parsing as it's currently written relies on the hash as a nice, easy-to-spot delimeter.

Reply:
Oh, sorry! I should have been more clear. I want to make sure `format!("{}{}", url, ":myfolder"))` _isn't_ parsed as a subdir. It's a suspicious input, where the implementation might plausibly fail, and I want to make sure that we interpret it correctly.

Does this make sense?